File: app/src/main/java/com/zacharee1/sliderpreferenceembedded/DiscreteSeekBarText.java
package com.zacharee1.sliderpreferenceembedded;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.adw.library.widgets.discreteseekbar.DiscreteSeekBar;

import java.text.NumberFormat;
import java.util.Locale;

public class DiscreteSeekBarText extends LinearLayout implements DiscreteSeekBar.OnProgressChangeListener {
    private DiscreteSeekBar mSeekBar;
    private TextView mTextView;
    private float mScale = 1.0F; //doesn't work for the popup view
    private int mPrecision = 2;

    private OnProgressChangeListener mListener;

    public DiscreteSeekBarText(Context context) {
        this(context, null);
    }

    public DiscreteSeekBarText(Context context, AttributeSet attrs) {
        this(context, attrs, R.attr.discreteSeekBarStyle);
    }

    public DiscreteSeekBarText(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        inflate(getContext(), R.layout.seekbar_with_text, this);

        mSeekBar = findViewById(R.id.seekbar);
        mTextView = findViewById(R.id.textview);

        mSeekBar.setOnProgressChangeListener(this);

        TypedValue typedValue = new TypedValue();
        Resources.Theme theme = getContext().getTheme();
        theme.resolveAttribute(R.attr.colorAccent, typedValue, true);
        int color = typedValue.data;

        mSeekBar.setThumbColor(color, color);
        mSeekBar.setScrubberColor(color);
        mSeekBar.setTrackColor(color);
        mSeekBar.setRippleColor(color);
    }

    @Override
    public void onStartTrackingTouch(DiscreteSeekBar seekBar) {
        if (mListener != null) mListener.onStartTrackingTouch(seekBar);
    }

    @Override
    public void onStopTrackingTouch(DiscreteSeekBar seekBar) {
        if (mListener != null) mListener.onStopTrackingTouch(seekBar);
    }

    @Override
    public void onProgressChanged(DiscreteSeekBar seekBar, int value, boolean fromUser) {
        if (mListener != null) mListener.onProgressChanged(seekBar, value, fromUser);

        setText(value);
    }

    public void setOnProgressChangeListener(OnProgressChangeListener listener) {
        mListener = listener;
    }

    public void setProgress(int progress) {
        mSeekBar.setProgress(progress);
        setText(progress);
    }

    public void setMin(int min) {
        mSeekBar.setMin(min);
    }

    public void setMax(int max) {
        mSeekBar.setMax(max);
    }

    public void setIndicatorFormatter(String formatter) {
        mSeekBar.setIndicatorFormatter(formatter);
    }

    public void setPopupIndicatorEnabled(boolean enabled) {
        mSeekBar.setIndicatorPopupEnabled(enabled);
    }

    public void setTextIndicatorEnabled(boolean enabled) {
        mTextView.setVisibility(enabled ? VISIBLE : GONE);
    }

    public void setScale(float scale) {
        mScale = scale;
    }

    public void setPrecision(int precision) {
        mPrecision = precision;
    }

    public DiscreteSeekBar getSeekBar() {
        return mSeekBar;
    }

    public int getProgress() {
        return mSeekBar.getProgress();
    }

    public int getMin() {
        return mSeekBar.getMin();
    }

    public int getMax() {
        return mSeekBar.getMax();
    }

    public boolean getPopupIndicatorEnabled() {
        return mSeekBar.getIndicatorPopupEnabled();
    }

    public boolean getTextIndicatorEnabled() {
        return mTextView.getVisibility() == VISIBLE;
    }

    private void setText(int text) {
        double scaled = text * mScale;

        String format = mSeekBar.getIndicatorFormatter();
        String floatFormat;

        if (scaled == (long) scaled) {
            floatFormat = String.format(Locale.US, "%d", (long)scaled);
        } else {
            floatFormat = String.format(Locale.US, "%s", scaled);
        }

        if (format == null) {
            mTextView.setText(mScale < 1 ? floatFormat : String.valueOf(scaled));
        } else {
            mTextView.setText(String.format(format, mScale < 1 ? floatFormat : String.valueOf(scaled)));
        }
    }

    public interface OnProgressChangeListener {
        void onProgressChanged(DiscreteSeekBar seekBar, int value, boolean fromUser);
        void onStopTrackingTouch(DiscreteSeekBar seekBar);
        void onStartTrackingTouch(DiscreteSeekBar seekBar);
    }
}

File: app/src/main/java/com/zacharee1/sliderpreferenceembedded/DiscreteSeekBarText.kt
package com.zacharee1.sliderpreferenceembedded

import android.content.Context
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import org.adw.library.widgets.discreteseekbar.DiscreteSeekBar
import java.util.*

class DiscreteSeekBarText @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = R.attr.discreteSeekBarStyle)
    : LinearLayout(context, attrs, defStyleAttr),
        DiscreteSeekBar.OnProgressChangeListener {

    private val seekBar: DiscreteSeekBar
    private val textView: TextView

    var scale = 1f //doesn't work for the popup view

    var listener: OnProgressChangeListener? = null

    var progress: Int
        get() = seekBar.progress
        set(progress) {
            seekBar.progress = progress
            text = progress.toString()
        }

    var min: Int
        get() = seekBar.min
        set(min) {
            seekBar.min = min
        }

    var max: Int
        get() = seekBar.max
        set(max) {
            seekBar.max = max
        }

    var popupIndicatorEnabled: Boolean
        get() = seekBar.indicatorPopupEnabled
        set(enabled) {
            seekBar.indicatorPopupEnabled = enabled
        }

    var textIndicatorEnabled: Boolean
        get() = textView.visibility == View.VISIBLE
        set(enabled) {
            textView.visibility = if (enabled) View.VISIBLE else View.GONE
        }

    var format: String
        get() = seekBar.indicatorFormatter
        set(value) {
            seekBar.indicatorFormatter = value
        }

    var text: String
        get() = textView.text.toString()
        set(value) {
            val scaled = (value.toInt() * scale).toDouble()

            val format = seekBar.indicatorFormatter
            val floatFormat: String

            floatFormat = if (scaled == scaled.toLong().toDouble()) {
                String.format(Locale.US, "%d", scaled.toLong())
            } else {
                String.format(Locale.US, "%.2f", scaled)
            }

            if (format == null) {
                textView.text = if (scale < 1) floatFormat else scaled.toString()
            } else {
                textView.text = String.format(format, if (scale < 1) floatFormat else scaled.toString())
            }
        }

    init {
        View.inflate(context, R.layout.seekbar_with_text, this)

        seekBar = findViewById(R.id.seekbar)
        textView = findViewById(R.id.textview)

        seekBar.setOnProgressChangeListener(this)

        val typedValue = TypedValue()
        val theme = context.theme
        theme.resolveAttribute(R.attr.colorAccent, typedValue, true)
        val color = typedValue.data

        seekBar.setThumbColor(color, color)
        seekBar.setScrubberColor(color)
        seekBar.setTrackColor(color)
        seekBar.setRippleColor(color)
    }

    override fun onStartTrackingTouch(seekBar: DiscreteSeekBar) {
        listener?.onStartTrackingTouch(seekBar)
    }

    override fun onStopTrackingTouch(seekBar: DiscreteSeekBar) {
        listener?.onStopTrackingTouch(seekBar)
    }

    override fun onProgressChanged(seekBar: DiscreteSeekBar, value: Int, fromUser: Boolean) {
        listener?.onProgressChanged(seekBar, value, fromUser)

        text = value.toString()
    }

    interface OnProgressChangeListener {
        fun onProgressChanged(seekBar: DiscreteSeekBar, value: Int, fromUser: Boolean)
        fun onStopTrackingTouch(seekBar: DiscreteSeekBar)
        fun onStartTrackingTouch(seekBar: DiscreteSeekBar)
    }
}
